staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go

```diff
@@ -0,0 +1,270 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	k8sscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/klog/v2"
+)
+
+// This handler serves the /apis endpoint for an aggregated list of
+// api resources indexed by their group version.
+type ResourceManager interface {
+	// Adds knowledge of the given groupversion to the discovery document
+	// If it was already being tracked, updates the stored APIVersionDiscovery
+	// Thread-safe
+	AddGroupVersion(groupName string, value apidiscoveryv2beta1.APIVersionDiscovery)
+
+	// Removes all group versions for a given group
+	// Thread-safe
+	RemoveGroup(groupName string)
+
+	// Removes a specific groupversion. If all versions of a group have been
+	// removed, then the entire group is unlisted.
+	// Thread-safe
+	RemoveGroupVersion(gv metav1.GroupVersion)
+
+	// Resets the manager's known list of group-versions and replaces them
+	// with the given groups
+	// Thread-Safe
+	SetGroups([]apidiscoveryv2beta1.APIGroupDiscovery)
+
+	http.Handler
+}
+
+type noopResourceManager struct{}
+
+func (noopResourceManager) AddGroupVersion(groupName string, value apidiscoveryv2beta1.APIVersionDiscovery) {
+}
+func (noopResourceManager) RemoveGroup(groupName string)                      {}
+func (noopResourceManager) RemoveGroupVersion(gv metav1.GroupVersion)         {}
+func (noopResourceManager) SetGroups([]apidiscoveryv2beta1.APIGroupDiscovery) {}
+func (noopResourceManager) WebService() *restful.WebService                   { return nil }
+func (noopResourceManager) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	resp.WriteHeader(http.StatusNotImplemented)
+}
+
+var _ ResourceManager = noopResourceManager{}
+
+func NewNoopResourceManager() ResourceManager {
+	return noopResourceManager{}
+}
+
+type resourceDiscoveryManager struct {
+	// Protects writes to all fields in struct
+	lock sync.RWMutex
+
+	// Writes protected by the lock. List if all apigroups & resources indexed
+	// by the resource manager
+	apiGroups map[string]*apidiscoveryv2beta1.APIGroupDiscovery
+
+	apiGroupNames []string
+
+	serializer runtime.NegotiatedSerializer
+	cachedGroupList
+}
+
+type cachedGroupList struct {
+	// Marks whether cache has been initialized
+	cachePopulated bool
+	// Cached ordered list of groups
+	groups []apidiscoveryv2beta1.APIGroupDiscovery
+}
+
+func NewResourceManager() ResourceManager {
+	scheme := runtime.NewScheme()
+	codecs := serializer.NewCodecFactory(scheme)
+	k8sscheme.AddToScheme(scheme)
+	apidiscoveryv2beta1.AddToScheme(scheme)
+	result := &resourceDiscoveryManager{serializer: codecs}
+	return result
+}
+
+func (rdm *resourceDiscoveryManager) SetGroups(groups []apidiscoveryv2beta1.APIGroupDiscovery) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	rdm.apiGroups = nil
+	rdm.apiGroupNames = nil
+	rdm.cachedGroupList = cachedGroupList{}
+
+	for _, group := range groups {
+		for _, version := range group.Versions {
+			rdm.addGroupVersionLocked(group.Name, version)
+		}
+	}
+}
```

Comment:
That looks like creating a new manager basically?

Reply:
it resets the data in the manager, do you have any thoughts on it?